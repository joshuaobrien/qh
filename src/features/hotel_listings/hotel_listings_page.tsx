import type React from "react";
import logo from "../../design_system/logo.png";
import { Body } from "../../design_system/typography";
import type { HotelService } from "../../services/hotel_service/hotel_service";
import styles from "./hotel_listing_page.module.css";
import { HotelListingsSummary } from "./hotel_listings_summary";
import { LoadingContent } from "./loading_content";
import { ReadyContent } from "./ready_content";
import { HotelListingsSkeleton } from "./skeleton/hotel_listings_skeleton";
import { HotelListingsSortOrderSelector } from "./sort_order_selector";
import { useHotelListings } from "./use_hotel_listings";

type HotelListingsPageProps = {
	hotelService: HotelService;
};

export const HotelListingsPage = ({ hotelService }: HotelListingsPageProps) => {
	const { sortOrder, setSortOrder, status, data, location, stayLengthNights } =
		useHotelListings(hotelService);

	const topLeft =
		status === "ready" ? (
			<HotelListingsSummary count={data.results.length} location={location} />
		) : undefined;

	const topRight = (
		<HotelListingsSortOrderSelector
			sortOrder={sortOrder}
			setSortOrder={setSortOrder}
		/>
	);

	// todo(josh): comprehensive components for the error case + empty response case
	let content: React.ReactNode;
	switch (status) {
		case "loading":
			content = <LoadingContent />;
			break;
		case "error":
			content = <Body>Sorry, there was a problem finding hotels.</Body>;
			break;
		case "ready":
			content =
				data.results.length === 0 ? (
					<Body>No results found.</Body>
				) : (
					<ReadyContent
						results={data.results}
						stayLengthNights={stayLengthNights}
					/>
				);
			break;
		default:
			status satisfies never;
	}

	return (
		<div className={styles.container}>
			<HotelListingsSkeleton
				logo={<img src={logo} alt="The Qantas logo" className={styles.logo} />}
				topLeft={topLeft}
				topRight={topRight}
				content={content}
			/>
		</div>
	);
};
