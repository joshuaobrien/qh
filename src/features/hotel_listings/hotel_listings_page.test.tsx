import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import {
	cleanup,
	render,
	screen,
	waitFor,
	within,
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";
import type {
	FindHotelsResponse,
	HotelService,
} from "../../services/hotel_service/hotel_service";
import { HotelListingsPage } from "./hotel_listings_page";

const queryClient = new QueryClient({
	defaultOptions: { queries: { retry: false } },
});

const mockHotelsData: FindHotelsResponse = {
	results: [
		{
			id: "1",
			property: {
				title: "Hotel Sydney",
				address: ["123 Main St", "Sydney"],
				previewImage: {
					url: "https://example.com/hotel_sydney.jpg",
					caption: "",
					imageType: "PRIMARY",
				},
				rating: { ratingValue: 4.5, ratingType: "star" },
				propertyId: "sydney-1",
			},
			offer: {
				name: "Deluxe Room",
				displayPrice: { amount: 200, currency: "AUD" },
				cancellationOption: { cancellationType: "FREE_CANCELLATION" },
				savings: { amount: 50, currency: "AUD" },
				promotion: {
					title: "",
					type: "MEMBER",
				},
			},
		},
		{
			id: "2",
			property: {
				title: "Luxury Melbourne",
				address: ["456 Collins St", "Melbourne"],
				previewImage: {
					url: "https://example.com/hotel_melbourne.jpg",
					caption: "",
					imageType: "PRIMARY",
				},
				rating: { ratingValue: 5, ratingType: "star" },
				propertyId: "melbourne-2",
			},
			offer: {
				name: "Executive Suite",
				displayPrice: { amount: 350, currency: "AUD" },
				cancellationOption: { cancellationType: "NOT_REFUNDABLE" },
				savings: { amount: 75, currency: "AUD" },
				promotion: {
					title: "",
					type: "MEMBER",
				},
			},
		},
	],
};

describe("HotelListingsPage", () => {
	let hotelService: HotelService;

	beforeEach(() => {
		hotelService = { findHotels: vi.fn() };
	});

	afterEach(() => {
		cleanup();
		queryClient.clear();
	});

	const renderWithQueryClient = (ui: React.ReactNode) =>
		render(
			<QueryClientProvider client={queryClient}>{ui}</QueryClientProvider>,
		);

	it("renders loading state", async () => {
		vi.mocked(hotelService).findHotels.mockReturnValue(new Promise(() => {}));
		renderWithQueryClient(<HotelListingsPage hotelService={hotelService} />);
		expect(screen.getByRole("list")).toBeDefined();
		expect(screen.getByLabelText("Loading")).toBeDefined();
	});

	it("renders error state", async () => {
		vi.mocked(hotelService).findHotels.mockRejectedValue(
			new Error("Failed to fetch hotels"),
		);
		renderWithQueryClient(<HotelListingsPage hotelService={hotelService} />);
		await waitFor(() => {
			expect(screen.queryByRole("list")).toBeNull();
			expect(
				screen.getByText("Sorry, there was a problem finding hotels."),
			).toBeDefined();
		});
	});

	it("renders hotel listings when data is available", async () => {
		vi.mocked(hotelService).findHotels.mockResolvedValue(mockHotelsData);
		renderWithQueryClient(<HotelListingsPage hotelService={hotelService} />);
		await waitFor(() => {
			expect(screen.getByText("Hotel Sydney")).toBeDefined();
			expect(screen.getByText("123 Main St, Sydney")).toBeDefined();
			expect(screen.getByText("Deluxe Room")).toBeDefined();
			expect(screen.getByText("Free cancellation")).toBeDefined();
			expect(screen.getByText("Save $50~")).toBeDefined();
		});
	});

	it("applies default sorting order on initial render", async () => {
		vi.mocked(hotelService).findHotels.mockResolvedValue(mockHotelsData);
		renderWithQueryClient(<HotelListingsPage hotelService={hotelService} />);
		expect(vi.mocked(hotelService).findHotels).toHaveBeenLastCalledWith({
			sortOrder: "high-first",
		});
	});

	it("changes sort order and verifies correct sorting", async () => {
		vi.mocked(hotelService).findHotels.mockResolvedValue(mockHotelsData);
		renderWithQueryClient(<HotelListingsPage hotelService={hotelService} />);
		expect(vi.mocked(hotelService).findHotels).toHaveBeenLastCalledWith({
			sortOrder: "high-first",
		});
		const select = screen.getByRole("combobox");
		await userEvent.selectOptions(select, "low-first");
		expect(vi.mocked(hotelService).findHotels).toHaveBeenLastCalledWith({
			sortOrder: "low-first",
		});
		expect(vi.mocked(hotelService.findHotels)).toHaveBeenCalledTimes(2);
	});

	it("only advertises free cancellation if there is a free cancellation offer", async () => {
		vi.mocked(hotelService).findHotels.mockResolvedValue(mockHotelsData);

		renderWithQueryClient(<HotelListingsPage hotelService={hotelService} />);

		const hotelSydneyElement = await screen.findByText("Hotel Sydney");

		const hotelSydneyListItem = hotelSydneyElement.closest("li");
		expect(hotelSydneyListItem).not.toBeNull();
		expect(
			// biome-ignore lint/style/noNonNullAssertion: there's a check above
			within(hotelSydneyListItem!).queryByText("Free cancellation"),
		).toBeDefined();
		expect(screen.getAllByText("Free cancellation")).toHaveLength(1);
	});

	it("matches snapshot for a loaded list of hotels", async () => {
		vi.mocked(hotelService).findHotels.mockResolvedValue(mockHotelsData);
		const { container } = renderWithQueryClient(
			<HotelListingsPage hotelService={hotelService} />,
		);
		await screen.findByText("Hotel Sydney");
		expect(container).toMatchSnapshot();
	});
});
