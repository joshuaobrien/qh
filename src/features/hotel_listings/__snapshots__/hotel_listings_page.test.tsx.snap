// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HotelListingsPage > matches snapshot for a loaded list of hotels 1`] = `
<div>
  <div
    class="_container_a5b42b"
  >
    <div
      class="_container_f57c3b"
    >
      <div
        class="_logo_f57c3b"
      >
        <img
          alt="The Qantas logo"
          class="_logo_a5b42b"
          src="/src/design_system/logo.png"
        />
      </div>
      <div
        class="_top_f57c3b"
      >
        <span
          class="_normal_ac58d8 _regular_ac58d8 _truncate_ac58d8"
          style="color: black;"
        >
          Showing 
          <span
            class="_normal_ac58d8 _semibold_ac58d8 _truncate_ac58d8"
            style="color: black;"
          >
            2
          </span>
           
          <span
            class="_normal_ac58d8 _italics_ac58d8 _regular_ac58d8 _truncate_ac58d8"
            style="color: black;"
          >
            hotels
             in
          </span>
           
          <span
            class="_normal_ac58d8 _semibold_ac58d8 _truncate_ac58d8"
            style="color: black;"
          >
            Sydney
          </span>
        </span>
        <div>
          <label
            for="sort-order"
          >
            Sort by 
          </label>
          <select
            id="sort-order"
          >
            <option
              value="high-first"
            >
              Price high-low
            </option>
            <option
              value="low-first"
            >
              Price low-high
            </option>
          </select>
        </div>
      </div>
      <ul>
        <li>
          <div
            class="_card_b6868b"
          >
            <div
              class="_imageContainer_b6868b"
            >
              <div
                class="_image_b6868b"
              >
                <img
                  aria-label=""
                  class="_container_4e0440"
                  src="https://example.com/hotel_sydney.jpg"
                />
              </div>
              <div
                class="_imageLabel_b6868b"
              >
                <span
                  class="_normal_ac58d8 _regular_ac58d8 _truncate_ac58d8"
                  style="color: red;"
                />
              </div>
            </div>
            <div
              class="_content_b6868b"
            >
              <div
                class="_upper_b6868b"
              >
                <div
                  class="_titleAndRatingContainer_b6868b"
                >
                  <div
                    class="_title_b6868b"
                  >
                    <h2
                      class="_large_ac58d8 _semibold_ac58d8 _truncate_ac58d8"
                      style="color: black;"
                    >
                      Hotel Sydney
                    </h2>
                  </div>
                  <div
                    class="_rating_b6868b"
                  >
                    <div
                      aria-label="Rating: 4.5 out of 5. (official rating)"
                      class="_container_b2fdaf"
                      role="img"
                    >
                      <span
                        aria-hidden="true"
                        class="_icon_b2fdaf"
                      >
                        <svg
                          fill="rgb(251, 203, 59)"
                          height="24"
                          viewBox="0 0 24 24"
                          width="24"
                        >
                          <title>
                            star
                          </title>
                          <path
                            d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                          />
                        </svg>
                      </span>
                      <span
                        aria-hidden="true"
                        class="_icon_b2fdaf"
                      >
                        <svg
                          fill="rgb(251, 203, 59)"
                          height="24"
                          viewBox="0 0 24 24"
                          width="24"
                        >
                          <title>
                            star
                          </title>
                          <path
                            d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                          />
                        </svg>
                      </span>
                      <span
                        aria-hidden="true"
                        class="_icon_b2fdaf"
                      >
                        <svg
                          fill="rgb(251, 203, 59)"
                          height="24"
                          viewBox="0 0 24 24"
                          width="24"
                        >
                          <title>
                            star
                          </title>
                          <path
                            d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                          />
                        </svg>
                      </span>
                      <span
                        aria-hidden="true"
                        class="_icon_b2fdaf"
                      >
                        <svg
                          fill="rgb(251, 203, 59)"
                          height="24"
                          viewBox="0 0 24 24"
                          width="24"
                        >
                          <title>
                            star
                          </title>
                          <path
                            d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                          />
                        </svg>
                      </span>
                      <span
                        aria-hidden="true"
                        class="_icon_b2fdaf"
                      >
                        <svg
                          fill="rgb(251, 203, 59)"
                          height="24"
                          viewBox="0 0 24 24"
                          width="24"
                        >
                          <title>
                            half star
                          </title>
                          <path
                            d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4V6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                          />
                        </svg>
                      </span>
                    </div>
                  </div>
                </div>
                <div
                  class="_address_b6868b"
                >
                  <h3
                    class="_normal_ac58d8 _regular_ac58d8 _truncate_ac58d8"
                    style="color: grey;"
                  >
                    123 Main St, Sydney
                  </h3>
                </div>
              </div>
              <div
                class="_lower_b6868b"
              >
                <div
                  class="_roomAndPromotionContainer_b6868b"
                >
                  <div
                    class="_room_b6868b"
                  >
                    <span
                      class="_normal_ac58d8 _regular_ac58d8 _truncate_ac58d8"
                      style="color: red;"
                    >
                      Deluxe Room
                    </span>
                  </div>
                  <div
                    class="_promotion_b6868b"
                  >
                    <span
                      class="_normal_ac58d8 _regular_ac58d8 _truncate_ac58d8"
                      style="color: green;"
                    >
                      Free cancellation
                    </span>
                  </div>
                </div>
                <div
                  class="_pricingContainer_b6868b"
                >
                  <div
                    class="_priceLabel_b6868b"
                  >
                    <span
                      class="_normal_ac58d8 _semibold_ac58d8 _truncate_ac58d8"
                      style="color: black;"
                    >
                      1
                    </span>
                    <span
                      class="_normal_ac58d8 _regular_ac58d8 _truncate_ac58d8"
                      style="color: black;"
                    >
                       night total (AUD)
                    </span>
                  </div>
                  <div
                    class="_price_b6868b"
                  >
                    <span>
                      <span
                        class="_extraLarge_ac58d8 _regular_ac58d8 _truncate_ac58d8"
                        style="color: black;"
                      >
                        200
                      </span>
                    </span>
                  </div>
                  <div
                    class="_savings_b6868b"
                  >
                    <span
                      class="_normal_ac58d8 _semibold_ac58d8 _truncate_ac58d8"
                      style="color: red;"
                    >
                      Save $50~
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </li>
        <li>
          <div
            class="_card_b6868b"
          >
            <div
              class="_imageContainer_b6868b"
            >
              <div
                class="_image_b6868b"
              >
                <img
                  aria-label=""
                  class="_container_4e0440"
                  src="https://example.com/hotel_melbourne.jpg"
                />
              </div>
              <div
                class="_imageLabel_b6868b"
              >
                <span
                  class="_normal_ac58d8 _regular_ac58d8 _truncate_ac58d8"
                  style="color: red;"
                />
              </div>
            </div>
            <div
              class="_content_b6868b"
            >
              <div
                class="_upper_b6868b"
              >
                <div
                  class="_titleAndRatingContainer_b6868b"
                >
                  <div
                    class="_title_b6868b"
                  >
                    <h2
                      class="_large_ac58d8 _semibold_ac58d8 _truncate_ac58d8"
                      style="color: black;"
                    >
                      Luxury Melbourne
                    </h2>
                  </div>
                  <div
                    class="_rating_b6868b"
                  >
                    <div
                      aria-label="Rating: 5 out of 5. (official rating)"
                      class="_container_b2fdaf"
                      role="img"
                    >
                      <span
                        aria-hidden="true"
                        class="_icon_b2fdaf"
                      >
                        <svg
                          fill="rgb(251, 203, 59)"
                          height="24"
                          viewBox="0 0 24 24"
                          width="24"
                        >
                          <title>
                            star
                          </title>
                          <path
                            d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                          />
                        </svg>
                      </span>
                      <span
                        aria-hidden="true"
                        class="_icon_b2fdaf"
                      >
                        <svg
                          fill="rgb(251, 203, 59)"
                          height="24"
                          viewBox="0 0 24 24"
                          width="24"
                        >
                          <title>
                            star
                          </title>
                          <path
                            d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                          />
                        </svg>
                      </span>
                      <span
                        aria-hidden="true"
                        class="_icon_b2fdaf"
                      >
                        <svg
                          fill="rgb(251, 203, 59)"
                          height="24"
                          viewBox="0 0 24 24"
                          width="24"
                        >
                          <title>
                            star
                          </title>
                          <path
                            d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                          />
                        </svg>
                      </span>
                      <span
                        aria-hidden="true"
                        class="_icon_b2fdaf"
                      >
                        <svg
                          fill="rgb(251, 203, 59)"
                          height="24"
                          viewBox="0 0 24 24"
                          width="24"
                        >
                          <title>
                            star
                          </title>
                          <path
                            d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                          />
                        </svg>
                      </span>
                      <span
                        aria-hidden="true"
                        class="_icon_b2fdaf"
                      >
                        <svg
                          fill="rgb(251, 203, 59)"
                          height="24"
                          viewBox="0 0 24 24"
                          width="24"
                        >
                          <title>
                            star
                          </title>
                          <path
                            d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                          />
                        </svg>
                      </span>
                    </div>
                  </div>
                </div>
                <div
                  class="_address_b6868b"
                >
                  <h3
                    class="_normal_ac58d8 _regular_ac58d8 _truncate_ac58d8"
                    style="color: grey;"
                  >
                    456 Collins St, Melbourne
                  </h3>
                </div>
              </div>
              <div
                class="_lower_b6868b"
              >
                <div
                  class="_roomAndPromotionContainer_b6868b"
                >
                  <div
                    class="_room_b6868b"
                  >
                    <span
                      class="_normal_ac58d8 _regular_ac58d8 _truncate_ac58d8"
                      style="color: red;"
                    >
                      Executive Suite
                    </span>
                  </div>
                  <div
                    class="_promotion_b6868b"
                  />
                </div>
                <div
                  class="_pricingContainer_b6868b"
                >
                  <div
                    class="_priceLabel_b6868b"
                  >
                    <span
                      class="_normal_ac58d8 _semibold_ac58d8 _truncate_ac58d8"
                      style="color: black;"
                    >
                      1
                    </span>
                    <span
                      class="_normal_ac58d8 _regular_ac58d8 _truncate_ac58d8"
                      style="color: black;"
                    >
                       night total (AUD)
                    </span>
                  </div>
                  <div
                    class="_price_b6868b"
                  >
                    <span>
                      <span
                        class="_extraLarge_ac58d8 _regular_ac58d8 _truncate_ac58d8"
                        style="color: black;"
                      >
                        350
                      </span>
                    </span>
                  </div>
                  <div
                    class="_savings_b6868b"
                  >
                    <span
                      class="_normal_ac58d8 _semibold_ac58d8 _truncate_ac58d8"
                      style="color: red;"
                    >
                      Save $75~
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
